PublicGitArchive/pga/cmd/cache.go

```diff
@@ -17,32 +18,84 @@ const (
 
 // updateCache checks whether a new version of the file in url exists and downloads it
 // to dest. It returns an error when it was not possible to update it.
-func updateCache(dest, source FileSystem, name string) error {
+func updateCache(ctx context.Context, dest, source FileSystem, name string) error {
 	logrus.Debugf("syncing %s to %s", source.Abs(name), dest.Abs(name))
 	if upToDate(dest, source, name) {
 		logrus.Debugf("local copy is up to date")
 		return nil
 	}
 
 	logrus.Debugf("local copy is outdated or non existent")
-	wc, err := dest.Create(name)
+	tmpName := name + ".tmp"
+	if err := copy(ctx, source, dest, name, tmpName); err != nil {
+		if cerr := dest.Remove(tmpName); cerr != nil {
+			logrus.Warningf("error removing temporary file %s: %v",
+				dest.Abs(tmpName), cerr)
+		}
+
+		return fmt.Errorf("could not copy to temporary file %s: %v",
+			dest.Abs(tmpName), err)
+	}
+
+	if err := dest.Rename(tmpName, name); err != nil {
+		return fmt.Errorf("rename %s to %s failed: %v",
+			dest.Abs(tmpName), dest.Abs(name), err)
+	}
+
+	return nil
+}
+
+func copy(ctx context.Context, source, dest FileSystem,
+	sourceName, destName string) (err error) {
+
+	wc, err := dest.Create(destName)
 	if err != nil {
-		return fmt.Errorf("could not create %s: %v", dest.Abs(name), err)
+		return fmt.Errorf("could not create %s: %v", dest.Abs(destName), err)
 	}
 
-	rc, err := source.Open(name)
+	rc, err := source.Open(sourceName)
 	if err != nil {
+		_ = wc.Close()
 		return err
 	}
-	defer rc.Close()
 
-	if _, err = io.Copy(wc, rc); err != nil {
-		return fmt.Errorf("could not copy %s to %s: %v", source.Abs(name), dest.Abs(name), err)
+	if _, err = cancelableCopy(ctx, wc, rc); err != nil {
+		_ = rc.Close()
+		_ = wc.Close()
+		return fmt.Errorf("could not copy %s to %s: %v",
+			source.Abs(sourceName), dest.Abs(destName), err)
+	}
+
+	if err := rc.Close(); err != nil {
+		_ = wc.Close()
+		return err
 	}
-	if err := wc.Close(); err != nil {
-		return fmt.Errorf("could not close %s: %v", dest.Abs(name), err)
+
+	return wc.Close()
+}
+
+const copyBufferSize = 512 * 1024
+
+func cancelableCopy(ctx context.Context, dst io.Writer, src io.Reader) (int64, error) {
```

Comment:
I'm almost tempted to propose this function to Go's stdlib.

Reply:
That would be great, but I doubt it gets accepted, since nothing in `io` or `io/ioutil` is cancellable. Also, there are prominent voices against that usage: [Context isn’t for cancellation](https://dave.cheney.net/2017/08/20/context-isnt-for-cancellation).